src/util/mod.rs

```diff
@@ -1,41 +1,15 @@
-use std::{io, slice};
+#[repr(C)] // guarantee 'bytes' comes after '_align'
+pub struct WorldAligned<Bytes: ?Sized> {
+    pub _align: [u32; 0],
+    pub bytes: Bytes,
+}
 
-// TODO: This is copy/pasted from gfx-hal, so we need to find a new place to put
-// this function
-pub fn read_spirv<R: io::Read + io::Seek>(mut x: R) -> io::Result<Vec<u32>> {
-    let size = x.seek(io::SeekFrom::End(0))?;
-    if size % 4 != 0 {
-        return Err(io::Error::new(
-            io::ErrorKind::InvalidData,
-            "input length not divisible by 4",
-        ));
-    }
-    if size > usize::max_value() as u64 {
-        return Err(io::Error::new(io::ErrorKind::InvalidData, "input too long"));
-    }
-    let words = (size / 4) as usize;
-    let mut result = Vec::<u32>::with_capacity(words);
-    x.seek(io::SeekFrom::Start(0))?;
-    unsafe {
-        // Writing all bytes through a pointer with less strict alignment when our type has no
-        // invalid bitpatterns is safe.
-        x.read_exact(slice::from_raw_parts_mut(
-            result.as_mut_ptr() as *mut u8,
-            words * 4,
-        ))?;
-        result.set_len(words);
-    }
+pub fn make_spirv<'a>(data: &'a [u8]) -> super::ShaderModuleSource<'a> {
     const MAGIC_NUMBER: u32 = 0x0723_0203;
-    if !result.is_empty() && result[0] == MAGIC_NUMBER.swap_bytes() {
-        for word in &mut result {
-            *word = word.swap_bytes();
-        }
-    }
-    if result.is_empty() || result[0] != MAGIC_NUMBER {
-        return Err(io::Error::new(
-            io::ErrorKind::InvalidData,
-            "input missing SPIR-V magic number",
-        ));
-    }
-    Ok(result)
+
+    let (pre, words, post) = unsafe { data.align_to::<u32>() };
+    assert_eq!(pre, &[], "data offset is not aligned to words!");
+    assert_eq!(post, &[], "data size is not aligned to words!");
+    assert_eq!(words[0], MAGIC_NUMBER, "wrong magic word {:x}!", words[0]);
```

Comment:
Can we extend this to: "wrong magic word {:x}. Make sure you are using a SPIRV file."

Reply:
"wrong magic word {:x}. Make sure you are using a **binary** SPIRV file."